### PR TITLE
Diagnostics refactor + missing but resolvable groupId fix

### LINKED ISSUE #48: Plugin configuration validation gives false-positives
There are a number of corner-cases that the [`PluginValidator`](https://github.com/eclipse/lemminx-maven/blob/master/lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java) is missing. Test's should be added for all these cases.

In my testing, the cases which provide a false positive are:
1. When a plugin's groupId is missing. This gets resolved when the groupId is added. 
Example test case: 
```XML
			<plugin>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.8.1</version>
				<configuration>
					<source>1.8</source>
					<target>1.8</target>
				</configuration>
			</plugin>
```
**Note:** the warning does not appear when only the groupId & artifactId is present (for this test case)

2. When certain goals are present (in this example, the `enforce` goal gives a warning, but not the plugin configuration).
Example test case: 
```XML
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-enforcer-plugin</artifactId>
				<version>3.0.0-M3</version>
				<executions>
					<execution>
						<id>enforce-maven</id>
						<goals>
							<goal>enforce</goal>
						</goals>
						<configuration>
							<rules>
								<requireMavenVersion>
									<version>3.0.5</version>
								</requireMavenVersion>
							</rules>
						</configuration>
					</execution>
				</executions>
			</plugin>
```
**Note:** The goal which is giving a warning is giving it's description on hover, which means it was correctly resolved from the plugin descriptor. 
image

3. When a plugin's pom is missing from the `.m2/repository/` folder. This might not be the actual cause, but in the below example case, maven-shade-plugin `3.2.2` & `3.2.1` give plugin configuration errors and aren't able to resolve plugin goals (no goal description on hover). The error is resolved when version `3.2.3` or `2.3` is used. Below is a comparison of the contents of `3.2.2` & `3.2.3`:
```bash
[aobuchow@localhost 3.2.2]$ pwd
/home/aobuchow/.m2/repository/org/apache/maven/plugins/maven-shade-plugin/3.2.2
[aobuchow@localhost 3.2.2]$ ls -la
total 148
drwxr-xr-x. 2 aobuchow aobuchow   4096 May  6 12:28 .
drwxrwxr-x. 6 aobuchow aobuchow   4096 May  6 14:33 ..
-rw-r--r--. 1 aobuchow aobuchow     95 May  6 14:32 m2e-lastUpdated.properties
-rw-r--r--. 1 aobuchow aobuchow 130877 May  6 12:28 maven-shade-plugin-3.2.2.jar
-rw-r--r--. 1 aobuchow aobuchow     40 May  6 12:28 maven-shade-plugin-3.2.2.jar.sha1
-rw-r--r--. 1 aobuchow aobuchow    178 May  6 12:28 _remote.repositories
```
```bash
[aobuchow@localhost 3.2.3]$ pwd
/home/aobuchow/.m2/repository/org/apache/maven/plugins/maven-shade-plugin/3.2.3
[aobuchow@localhost 3.2.3]$ ls -la
total 168
drwxr-xr-x. 2 aobuchow aobuchow   4096 May  6 14:33 .
drwxrwxr-x. 6 aobuchow aobuchow   4096 May  6 14:33 ..
-rw-r--r--. 1 aobuchow aobuchow     95 May  6 14:34 m2e-lastUpdated.properties
-rw-r--r--. 1 aobuchow aobuchow 132252 May  6 14:33 maven-shade-plugin-3.2.3.jar
-rw-r--r--. 1 aobuchow aobuchow     40 May  6 14:33 maven-shade-plugin-3.2.3.jar.sha1
-rw-r--r--. 1 aobuchow aobuchow  10420 May  6 14:33 maven-shade-plugin-3.2.3.pom
-rw-r--r--. 1 aobuchow aobuchow     40 May  6 14:33 maven-shade-plugin-3.2.3.pom.sha1
-rw-r--r--. 1 aobuchow aobuchow    216 May  6 14:33 _remote.repositories
```
To briefly summarise this case: the plugin descriptor isin't being resolved for `maven-shade-plugin 3.2.2` (error on goals & configuration), but is being resolved for `maven-shade-plugin 3.2.3`.
Example test case:
```XML
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-shade-plugin</artifactId>
				<version>3.2.2</version>
				<executions>
					<execution>
						<phase>package</phase>
						<goals>
							<goal>shade</goal>
						</goals>
						<configuration>
							<transformers>
								<transformer
									implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
									<mainClass>org.eclipse.dash.licenses.cli.Main</mainClass>
								</transformer>
							</transformers>
							<createDependencyReducedPom>
								false
							</createDependencyReducedPom>
						</configuration>
					</execution>
				</executions>
			</plugin>
```

### LINKED ISSUE #74: Check Maven Project/Model when plugin is unresolvable
In certain cases, a plugin's groupId or version tag may be missing in a given pom, but the plugin can still be resolved. In this case, the Maven Project or Model should be checked to see if it contains the plugin that is being resolved. 

Example test case:
```XML
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>demo</groupId>
  <artifactId>demo</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <build>
    <plugins>
      <plugin>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.8.1</version>
        <configuration>
          <release>14</release>
          <compilerArgs>
            <arg>--enable-preview</arg>
          </compilerArgs>
        </configuration>
      </plugin>
    </plugins>
  </build>
</project>
```

Related m2e issue: https://bugs.eclipse.org/bugs/show_bug.cgi?id=563624

### PR DESCRIPTION
**Pull request by AObuchow**
Reenable  goal diagnostics

Test false goal diagnostics (part of #48)

If the groupId of a plugin is missing but its artifactId is present,
search the projects plugins for it in order to resolve the plugin (part
of #74)

Fix #48
Fix #74
Signed-off-by: Andrew Obuchowicz <aobuchow@redhat.com>

### CODE REVIEW

File: lemminx-maven/pom.xml [hunk 1]
```diff
@@ -115,7 +115,7 @@
 		<groupId>org.apache.maven.plugins</groupId>
 		<artifactId>maven-shade-plugin</artifactId>
 		<version>3.2.2</version>
-		<type>pom</type>
+		<type>test-jar</type>
```
Comment (AObuchow) [inline lemminx-maven/pom.xml:118]: Is this necessary? Double-check if `jar` will suffice.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/DiagnosticRequest.java [hunk 2]
```diff
@@ -62,6 +83,16 @@ public Range getRange() {
 				((DOMElement) node).getEndTagOpenOffset(), xmlDocument);
 	}
 
+	public void addDiagnostic(String errorMessage, DiagnosticSeverity severity) {
+		Diagnostic diagnostic = new Diagnostic(this.getRange(), errorMessage, severity,
+				this.getXMLDocument().getDocumentURI(), "XML");
+
+		// TODO: Should diagnostics just be a set?
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/DiagnosticRequest.java:90]: Do this `TODO`?

Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/DiagnosticRequest.java:90]: Yes, probably. Especially since they're actually not ordered.

Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/DiagnosticRequest.java:90]: Sounds good, will do.

**Comment by AObuchow** [comment]
Maybe also assert that the range of the diagnostic for unresolvable plugins is correct (in the tests).

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java [hunk 3]
```diff
@@ -82,80 +76,66 @@ public Diagnostic validateConfiguration(DiagnosticRequest diagnosticRequest) {
...
 					| InvalidPluginDescriptorException e) {
 				e.printStackTrace();
 				// A diagnostic was already added in validatePluginResolution()
 			}
 		}
-		return null;
 	}
 
-	private static Diagnostic internalValidateGoal(DiagnosticRequest diagnosticReq, PluginDescriptor pluginDescriptor) {
+	private static void internalValidateGoal(DiagnosticRequest diagnosticReq, PluginDescriptor pluginDescriptor) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java:112]: Functions (which don't have side effects) usually leads to easier maintenance than methods like this one that are built on side effects.
So I'd suggest the method becomes something like `private Optional<Diagnostic> validateGoal(DOMNode requestNode, descriptor)`, so consumers could just invoke `validateGoal(node, descriptor).ifPresent(request::addDiagnostic)`. This would also allow easier fine grain testing if necessary.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java [hunk 4]
```diff
@@ -82,80 +76,66 @@ public Diagnostic validateConfiguration(DiagnosticRequest diagnosticRequest) {
...
-
-		return new Diagnostic(diagnosticReq.getRange(), "Invalid goal for this plugin: " + goal,
-				DiagnosticSeverity.Warning, diagnosticReq.getXMLDocument().getDocumentURI(), "XML");
-
+		diagnosticReq.addDiagnostic("Invalid goal for this plugin: " + goal, DiagnosticSeverity.Warning);
 	}
 
-	private static Diagnostic internalValidateConfiguration(DiagnosticRequest diagnosticReq,
-			List<Parameter> parameters) {
+	private static void internalValidateConfiguration(DiagnosticRequest diagnosticReq, List<Parameter> parameters) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java:127]: same as above

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java [hunk 5]
```diff
@@ -82,80 +76,66 @@ public Diagnostic validateConfiguration(DiagnosticRequest diagnosticRequest) {
...
-				}
+				DiagnosticRequest childDiagnosticReq = new DiagnosticRequest(childNode, diagnosticRequest);
+				internalValidateConfiguration(childDiagnosticReq, parameters);
 			}
 		}
-		return null;
 	}
 
-	public Diagnostic validateGoal(DiagnosticRequest diagnosticRequest) {
+	public void validateGoal(DiagnosticRequest diagnosticRequest) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java:89]: Functions (which don't have side effects) usually leads to easier maintenance than methods like this one that are built on side effects.
So I'd suggest the method becomes something like private Optional<Diagnostic> validateGoal(DOMNode requestNode, descriptor), so consumers could just invoke validateGoal(node, descriptor).ifPresent(request::addDiagnostic). This would also allow easier fine grain testing if necessary.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java [hunk 6]
```diff
@@ -35,44 +34,39 @@ public PluginValidator(MavenProjectCache cache, RepositorySystemSession repoSess
...
+								+ MavenPluginUtils.LINE_BREAK + "Additional information: " + errorMessage,
+						DiagnosticSeverity.Warning);
 			});
 			return false;
 		}
 		return true;
 	}
 
-	public Diagnostic validateConfiguration(DiagnosticRequest diagnosticRequest) {
+	public void validateConfiguration(DiagnosticRequest diagnosticRequest) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java:61]: See below, about returning values instead of using side effects on parameters.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenDiagnosticParticipant.java [hunk 7]
```diff
@@ -50,29 +51,17 @@ public void doDiagnostics(DOMDocument xmlDocument, List<Diagnostic> diagnostics,
 		
 		projectCache.getProblemsFor(xmlDocument).stream().map(this::toDiagnostic).forEach(diagnostics::add);
 		DOMElement documentElement = xmlDocument.getDocumentElement();
-		HashMap<String, Function<DiagnosticRequest, Diagnostic>> tagDiagnostics = configureDiagnosticFunctions(
-				xmlDocument);
+		HashMap<String, Consumer<DiagnosticRequest>> tagDiagnostics = configureDiagnosticFunctions();
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenDiagnosticParticipant.java:54]: See comments below, I think Function drive to better code than consumers.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java (hunk 3 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java:112]: +1 I knew something felt wrong about how the diagnostics were being added.

Will make the fix + other related fixes.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPluginUtils.java [hunk 8]
```diff
@@ -94,14 +96,21 @@ public static PluginDescriptor getContainingPluginDescriptor(IPositionRequest re
...
 		Plugin plugin = project.getPlugin(pluginKey);
-		if (plugin == null) {
+		if (plugin == null && project.getPluginManagement() != null) {
 			plugin = project.getPluginManagement().getPluginsAsMap().get(pluginKey);
+			
+			if (plugin == null && artifactId.isPresent()) {
+				//pluginArtifactMap will be empty if PluginManagement is null
+				for (Entry <String, Artifact> entry : project.getPluginArtifactMap().entrySet() ) {
+					if (entry.getValue().getArtifactId().equals(artifactId.get())) {
+						plugin = project.getPlugin(entry.getKey());
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPluginUtils.java:106]: This is to fix #74

File: lemminx-maven/src/test/java/org/eclipse/lemminx/maven/test/LocalPluginTest.java [hunk 9]
```diff
@@ -136,16 +137,28 @@ public void testPluginDiagnosticMissingPlugin()
 		DOMDocument document = createDOMDocument("/pom-plugin-diagnostic-missing-plugin.xml", languageService);
 		System.out.println(languageService.doDiagnostics(document, () -> {
```
Comment (AObuchow) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/maven/test/LocalPluginTest.java:138]: remove system.out

**Comment by AObuchow** [comment]
@mickaelistria this should be good to review now. Sorry for the kind of large change (addressing two bugs at once).

File: lemminx-maven/src/test/java/org/eclipse/lemminx/maven/test/LocalPluginTest.java [hunk 10]
```diff
@@ -136,16 +137,29 @@ public void testPluginDiagnosticMissingPlugin()
...
-				"Incomplete GAV")));
+		}, new XMLValidationSettings()).size() == 1); 
+	}
+	
+	@Test
+	public void testDiagnosticMissingGroupId() throws IOException, URISyntaxException {
+		DOMDocument document = createDOMDocument("/pom-plugin-diagnostic-missing-groupid.xml", languageService);
+		List<Diagnostic> diagnostics = languageService.doDiagnostics(document, () -> {
+		}, new XMLValidationSettings());
+		System.out.println(diagnostics);
```
Comment (AObuchow) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/maven/test/LocalPluginTest.java:153]: Remove before merge.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java [hunk 11]
```diff
@@ -82,80 +79,73 @@ public Diagnostic validateConfiguration(DiagnosticRequest diagnosticRequest) {
 			// A diagnostic was already added in validatePluginResolution()
 		}
 		if (parameters.isEmpty()) {
-			return null;
+			return Optional.ofNullable(null);
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java:82]: Wouldn't `Optional.empty()` be better here?

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java [hunk 12]
```diff
@@ -82,80 +79,73 @@ public Diagnostic validateConfiguration(DiagnosticRequest diagnosticRequest) {
...
+		
+		return Optional.of(diagnostics);
 	}
 
-	public Diagnostic validateGoal(DiagnosticRequest diagnosticRequest) {
+	public Optional<List<Diagnostic>> validateGoal(DiagnosticRequest diagnosticRequest) {
 		DOMNode node = diagnosticRequest.getNode();
 		if (node == null) {
-			return null;
+			return Optional.ofNullable(null);
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java:97]: `Optional.empty()` ?

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java [hunk 13]
```diff
@@ -34,21 +32,15 @@ public void setPomFile(File pomFile) throws FileNotFoundException, IOException,
 
 	public Diagnostic validateSubModuleExistence(DiagnosticRequest diagnosticRequest) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java:33]: This could also be an `Optional<Diagnostic>`, or even an `Optional<List<Diagnostic>>` so this could be referenced in MavenDiagnosticParticipant.tagDiagnostics.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java [hunk 14]
```diff
@@ -35,44 +35,41 @@ public PluginValidator(MavenProjectCache cache, RepositorySystemSession repoSess
 		this.repoSession = repoSession;
 		this.pluginManager = pluginManager;
 	}
-	
-	public boolean validatePluginResolution(DiagnosticRequest diagnosticRequest) {
+
+	public Diagnostic validatePluginResolution(DiagnosticRequest diagnosticRequest) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java:39]: This could also be an `Optional<Diagnostic>`, or even an `Optional<List<Diagnostic>>` so this could be referenced in MavenDiagnosticParticipant.tagDiagnostics.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/VersionValidator.java [hunk 15]
```diff
@@ -12,34 +12,24 @@
...
 import org.apache.maven.model.Dependency;
-import org.eclipse.lemminx.dom.DOMDocument;
 import org.eclipse.lemminx.dom.DOMNode;
 import org.eclipse.lsp4j.Diagnostic;
 import org.eclipse.lsp4j.DiagnosticSeverity;
-import org.eclipse.lsp4j.Range;
 
 public class VersionValidator {
 
 	public static Diagnostic validateVersion(DiagnosticRequest diagnosticRequest) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/VersionValidator.java:21]: This could also be an `Optional<Diagnostic>`, or even an `Optional<List<Diagnostic>>` so this could be referenced in MavenDiagnosticParticipant.tagDiagnostics.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java (hunk 14 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java]: +1

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java (hunk 11 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PluginValidator.java]: Yes my bad :P

**Comment by AObuchow** [comment]
@mickaelistria thanks, I addressed your comments.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java [hunk 16]
```diff
@@ -32,23 +33,19 @@ public void setPomFile(File pomFile) throws FileNotFoundException, IOException,
...
 		String tagContent = null;
 		if (node.hasChildNodes()) {
 			tagContent = node.getChild(0).getNodeValue(); // tagContent is the module to validate eg.
 															// <module>tagContent</module>
 		}
 		if (node.hasChildNodes() && !model.getModules().contains(tagContent)) {
-			diagnostic = new Diagnostic(range, String.format("Module '%s' does not exist", tagContent),
-					DiagnosticSeverity.Error, xmlDocument.getDocumentURI(), "XML");
-			diagnosticRequest.getDiagnostics().add(diagnostic);
+			List<Diagnostic> diagnostics = new ArrayList<>();
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java:44]: You can use `Collections.singletonList(diagnosticRequest.createDiagnostic(....))` instead of manually instantiating a new ArrayList. Singleton collections will consume slightly less memory.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java [hunk 17]
```diff
@@ -32,23 +33,19 @@ public void setPomFile(File pomFile) throws FileNotFoundException, IOException,
 		model = mavenreader.read(new FileReader(pomFile));
 	}
 
-	public Diagnostic validateSubModuleExistence(DiagnosticRequest diagnosticRequest) {
+	public Optional<Object> validateSubModuleExistence(DiagnosticRequest diagnosticRequest) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java:36]: Why not `Optional<List<Diagnostic>>` here?

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/VersionValidator.java [hunk 18]
```diff
@@ -8,38 +8,34 @@
...
 		Artifact artifact = null;
-		Range range = diagnosticRequest.getRange();
-		Diagnostic diagnostic = null;
 		try {
 			 artifact = new DefaultArtifact(model.getGroupId(), model.getArtifactId(), model.getVersion(), model.getScope(), model.getType(), model.getClassifier(), new DefaultArtifactHandler(model.getType()));
 			 if (!artifact.isSelectedVersionKnown()) {
-				diagnostic = new Diagnostic(range, "Version Error", DiagnosticSeverity.Error,
-						xmlDocument.getDocumentURI(), "XML");
-				diagnosticRequest.getDiagnostics().add(diagnostic);
+				 List<Diagnostic> diagnostics = new ArrayList<>();
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/VersionValidator.java:32]: Same as above, you can probably use `Collections.singletonList()`

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java (hunk 16 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java:44]: Thanks I've wondered what was the nicer way of doing that. Will adjust.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java (hunk 17 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/SubModuleValidator.java:36]: Oops, good catch.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/VersionValidator.java (hunk 18 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/VersionValidator.java]: > Some more minor comments.

Just addressed these.

> going functional here is a cool pattern and keeping a node-name/functon registry leads to nice code, we may consider this for other classes later.

Agreed, it's something to consider.

**Comment by AObuchow** [comment]
>Some more minor comments.

Just addressed these.

> going functional here is a cool pattern and keeping a node-name/functon registry leads to nice code, we may consider this for other classes later.

Agreed, it's something to consider.

**Comment by AObuchow** [comment]
Should be good now.

**Comment by mickaelistria** [comment]
Looks good to merge IMO.
